dashboard/src/app/workspace/share-workspace/share-workspace.controller.js

```diff
@@ -0,0 +1,266 @@
+    this.users = [];
+
+    permissions.forEach((permission) => {
+      let userId = permission.user;
+      let user = this.codenvyUser.getUserFromId(userId);
+      if (user) {
+        this.formUserItem(user, permission);
+      } else {
+        this.codenvyUser.fetchUserId(userId).then(() => {
+          this.formUserItem(this.codenvyUser.getUserFromId(userId), permission);
+        });
+      }
+    });
+  }
+
+  /**
+   * Forms item to display with permissions and user data.
+   *
+   * @param user user data
+   * @param permissions permissions data
+   */
+  formUserItem(user, permissions) {
+    let userItem = angular.copy(user);
+    userItem.permissions = permissions;
+    this.users.push(userItem);
+  }
+
+  shareWorkspace() {
+    let permissionPromises = [];
+    this.recipeId = this.getRecipeId();
+
+    this.existingUsers.forEach((user) => {
+      permissionPromises.push(this.storeWorkspacePermissions(user));
+      permissionPromises.push(this.storeRecipePermissions(user));
+    });
+
+    this.$q.all(permissionPromises).then(() => {
+      //Clear share input data:
+      this.emails.length = 0;
+      this.existingUsers.clear();
+
+      this.refreshWorkspacePermissions();
+    }, (error) => {
+      this.isLoading = false;
+      this.cheNotification.showError(error.data && error.data.message ? error.data.message : 'Failed to share workspace.');
+    });
+  }
+
+  /**
+   * Stores user premissions of the current workspace.
+   *
+   * @param user user data
+   * @returns {*} promise with store permissions request
+   */
+  storeWorkspacePermissions(user) {
+    let permission = {};
+    permission.user = user.id;
+    permission.domain = this.workspaceDomain;
+    permission.instance = this.workspaceId;
+    permission.actions = this.actions;
+
+    return this.codenvyPermissions.storePermissions(permission);
+  }
+
+  /**
+   * Strores user permissions for current workspace's recipe.
+   *
+   * @param user user data
+   * @returns {*} promise with store permissions request
+   */
+  storeRecipePermissions(user) {
+    let permission = {};
+    permission.user = user.id;
+    permission.domain = this.recipeDomain;
+    permission.instance = this.recipeId;
+    permission.actions = ['read'];
+
+    return this.codenvyPermissions.storePermissions(permission);
+  }
+
+  /**
+   * Removes user permissions for current workspace
+   *
+   * @param user user
+   */
+  removePermissions(user) {
+    this.isLoading = true;
+    this.codenvyPermissions.removeWorkspacePermissions(user.permissions.instance, user.id).then(() => {
+      this.refreshWorkspacePermissions();
+    }, (error) => {
+      this.isLoading = false;
+      this.cheNotification.showError(error.data && error.data.message ? error.data.message : 'Failed to remove user ' + user.email + ' permissions.');
+    });
+  }
+
+  /**
+   * Find user by its email.
+   *
+   * @param email user's email
+   * @returns {*}
```

Comment:
it doesn't seem clear to me that when we find a user, we're returning here the email (and not the 'user' object)